### ISSUE TITLE
STK_mesh_base Mac build errors

### ISSUE THREAD

**Issue by spdomin** [trilinos/Trilinos#13324]
## Bug Report
@alanw0, after a hiatus in supporting nightly Mac builds, I noticed a few build errors in stk_mesh_base. I have not bisected since it should be simple to resolve.

### Description
[ 79%] Building CXX object packages/stk/stk_mesh/stk_mesh/base/CMakeFiles/stk_mesh_base.dir/__/baseImpl/DeletedEntityCache.cpp.o
In file included from /Users/naluIt/gitHubWork/nightlyBuildAndTest/Trilinos/packages/stk/stk_mesh/stk_mesh/baseImpl/DeletedEntityCache.cpp:2:
In file included from /Users/naluIt/gitHubWork/nightlyBuildAndTest/Trilinos/packages/stk/stk_mesh/stk_mesh/baseImpl/MeshModification.hpp:44:
/Users/naluIt/gitHubWork/nightlyBuildAndTest/Trilinos/packages/stk/stk_mesh/stk_mesh/baseImpl/DeletedEntityCache.hpp:47:14: error: no
      template named 'unordered_map' in namespace 'std'
typedef std::unordered_map<EntityKey, Entity::entity_value_type, std::hash<EntityKey>> GhostReuseMap;
        ~~~~~^
/Users/naluIt/gitHubWork/nightlyBuildAndTest/Trilinos/packages/stk/stk_mesh/stk_mesh/baseImpl/DeletedEntityCache.cpp:14:22: error: type
      'stk::mesh::GhostReuseMap' (aka 'int') does not provide a subscript operator
    m_ghost_reuse_map[m_bulkData.entity_key(entity)] = entity.local_offset();
    ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/naluIt/gitHubWork/nightlyBuildAndTest/Trilinos/packages/stk/stk_mesh/stk_mesh/baseImpl/DeletedEntityCache.cpp:40:26: error: 
      invalid range expression of type 'int'; no viable 'begin' function available
  for (auto keyAndOffset : m_ghost_reuse_map) {
                         ^ ~~~~~~~~~~~~~~~~~
/Users/naluIt/gitHubWork/nightlyBuildAndTest/Trilinos/packages/stk/stk_mesh/stk_mesh/baseImpl/DeletedEntityCache.cpp:43:20: error: member
      reference base type 'stk::mesh::GhostReuseMap' (aka 'int') is not a structure or union
  m_ghost_reuse_map.clear();
### Steps to Reproduce
commit f70ac4d14e3b37e0f16e52d8095bf4b9a97da549
Merge: 24eb97a b0895ea

**Comment by alanw0** [comment]
Thanks for the report, this is a clear case of a missing include.
I'm surprised it works on other compilers, I guess there are indirect includes somehow getting it.
I'll get an update in as soon as I can.

**Comment by alanw0** [comment]
> Looks like this attempt to update STK failed.

Yeah we've had some cmake/macro changes recently related to the sierra cmake conversion. The trilinos PR testing caught a unit-test failure that I'm fixing now.
I guess it will notify you when it succeeds.

**Comment by spdomin** [comment]
I am now noting hangs on the new nightly process (non-Mac, simply Linux).

I will perform a bisect and see what is going on.

**Comment by alanw0** [comment]
> I am now noting hangs on the new nightly process (non-Mac, simply Linux).
> 
> I will perform a bisect and see what is going on.

Ugh. Let me know what you find.

**Comment by spdomin** [comment]
At least we have a small unit test that is showing the hang: Hex8Mesh.faceBasic

I have a couple more iterations. There was also a Kokkos snapshot...

**Comment by ndellingwood** [comment]
@spdomin If the test failure is correlated with the kokkos snapshot the hang could indicate something like a View creation or destruction within a parallel region (failure would for example be similar to https://github.com/trilinos/Trilinos/issues/13328 ). If that is the case, there is a new tool being developed that is helpful for detecting those cases, https://github.com/kokkos/kokkos-tools/pull/267 , running the hanging test with that tool loaded would help hone in on locations of code to inspect

**Comment by spdomin** [comment]
Great, I am on the final iteration:

commit f8ff2ad41462ea8af664241df5044928799e5984 (HEAD)
Author: Nathan Ellingwood <ndellin@sandia.gov>
Date:   Wed Aug 7 16:39:21 2024 -0600

    stk: modify test to prevent allocation in parallel region
    
    modify NgpMeshTest.volatileFastSharedCommMap to prevent allocation in a parallel region, which can result in deadlock with kokkos version 4.4
    address issue #13328
    
    Co-authored-by: Christian Trott <crtrott@sandia.gov>
    Signed-off-by: Nathan Ellingwood <ndellin@sandia.gov>

and this code base actually does not build cleanly...

[ 45%] Building CXX object packages/kokkos/containers/src/CMakeFiles/kokkoscontainers.dir/impl/Kokkos_UnorderedMap_impl.cpp.o
In file included from /fgs/spdomin/nightly/Trilinos/packages/kokkos/core/src/View/MDSpan/Kokkos_MDSpan_Extents.hpp:25,
                 from /fgs/spdomin/nightly/Trilinos/packages/kokkos/core/src/Kokkos_View.hpp:40,
                 from /fgs/spdomin/nightly/Trilinos/packages/kokkos/core/src/Kokkos_Parallel.hpp:31,
                 from /fgs/spdomin/nightly/Trilinos/packages/kokkos/core/src/Kokkos_MemoryPool.hpp:26,
                 from /fgs/spdomin/nightly/Trilinos/packages/kokkos/core/src/Kokkos_TaskScheduler.hpp:34,
                 from /fgs/spdomin/nightly/Trilinos/packages/kokkos/core/src/Serial/Kokkos_Serial.hpp:37,
                 from /fgs/spdomin/nightly/Trilinos/packages/kokkos/core/src/decl/Kokkos_Declare_SERIAL.hpp:21,
                 from /fgs/spdomin/nightly/Trilinos/build_nightly_release_10.3.0/packages/kokkos/KokkosCore_Config_DeclareBackend.hpp:22,
                 from /fgs/spdomin/nightly/Trilinos/packages/kokkos/core/src/Kokkos_Core.hpp:45,
                 from /fgs/spdomin/nightly/Trilinos/packages/kokkos/containers/src/Kokkos_UnorderedMap.hpp:30,
                 from /fgs/spdomin/nightly/Trilinos/packages/kokkos/containers/src/impl/Kokkos_UnorderedMap_impl.cpp:21:
/fgs/spdomin/nightly/Trilinos/packages/kokkos/core/src/View/MDSpan/Kokkos_MDSpan_Header.hpp:47:10: fatal error: mdspan/mdspan.hpp: No such file or directory
   47 | #include <mdspan/mdspan.hpp>

@crtrott or @alanw0 -  how shall I proceed to bisect the new Nalu hang?

**Comment by spdomin** [comment]
I will create a new issue

**Comment by spdomin** [comment]
@alanw0 - STK is off the hook:) See: https://github.com/trilinos/Trilinos/issues/13351

I may need to chat with you about this view-of-views thing... It reminds me of my photo-art that I often title "View from the [flower | stream | etc]

**Comment by alanw0** [comment]
> @alanw0 - STK is off the hook:) See: #13351
> 
> I may need to chat with you about this view-of-views thing... It reminds me of my photo-art that I often title "View from the [flower | stream | etc]

I always like being off the hook.
But seriously, I'm pretty sure nalu doesn't use the stk class that was the subject of that issue 13328 mentioned above. That class (NgpMesh/HostMesh) is related to NGP code.

**Comment by spdomin** [comment]
Looks like the Mac build and test is back online - thanks. I see the same hang that I will report on: https://github.com/trilinos/Trilinos/issues/13351
